src/webgpu/api/operation/render_pipeline/primitive_topology.spec.ts

```diff
@@ -0,0 +1,376 @@
+export const description = `Test primitive topology rendering.
+
+Draw a primitive using 6 vertices with each topology and check if the pixel is covered.
+
+Vertex sequence and coordinates are the same for each topology:
+  - Vertex buffer = [v1, v2, v3, v4, v5, v6]
+  - Topology = [point-list, line-list, line-strip, triangle-list, triangle-strip]
+
+Test locations are framebuffer coordinates:
+  - Pixel value { valid: green, invalid: black, format: 'rgba8unorm'}
+  - Test point is valid if the pixel value equals the covered pixel value at the test location.
+
+  Topology: point-list    Test points(s)
+
+  v2      v3      v6      Every vertex.
+
+
+
+  v1      v4      v5
+
+  Topology: line-list (3 lines)
+
+  v2       v4      v6
+  |        |       |      Center of each vertical line segment.
+  |        |       |
+  |        |       |
+  v1       v3      v5
+
+  Topology: line-strip (5 lines)
+
+  v2 ----- v3      v6
+  |        |       |        Center of each vertical line segment.
+  |        |       |      + Center of each horizontal line segment.
+  |        |       |
+  v1       v4 ---- v5
+
+  Topology: triangle-list (2 triangles)
+
+  v2 ----- v3      v6
+  | xxx           x |     Center of two triangle(s): {v1,v2,v3} and {v4,v5,v6}.
+  | xx           xx |
+  | x           xxx |
+  v1       v4 ---- v5
+
+  Topology: triangle-strip (4 triangles)
+
+  v2 ----- v3       v6
+  | xxx xxx x     x |       Center of two triangle(s): {v1,v2,v3} and {v4,v5,v6}.
+  | xx   xx xx   xx |     + Center of two triangle(s): {v2,v3,v4} and {v3,v4,v5}.
+  | x     x xxx xxx |
+  v1        v4 ---- v5
```

Comment:
The triangles v1-v2-v3 and v2-v3-v4 overlap (though it's hard to show that in the diagram).

Couldn't it happen that the centroid of one is within the other, so testing the centroids doesn't guarantee that the triangle was rasterized? (It depends on the exact triangle shapes, but I didn't look closely enough at them to be sure. Would be ideal if this problem were avoided by construction, e.g. avoiding overlapping triangles.)

Reply:
Good point, staggered the points.